ADALiOS/ADALiOS/ADWebAuthController.m

```diff
@@ -111,6 +112,104 @@ - (id)init
     return self;
 }
 
+- (void)cancelCurrentWebAuthSession
+{
+    [self webAuthenticationDidCancel];
+}
+
+- (BOOL)cancelCurrentWebAuthSessionWithError:(ADAuthenticationError*)error
+{
+    return [self endWebAuthenticationWithError:error orURL:nil];
+}
+
+#pragma mark - Private Methods
+
+- (void)dispatchCompletionBlock:(ADAuthenticationError *)error URL:(NSURL *)url
+{
+    // NOTE: It is possible that race between a successful completion
+    //       and the user cancelling the authentication dialog can
+    //       occur causing this method to be called twice. The race
+    //       cannot be blocked at its root, and so this method must
+    //       be resilient to this condition and should not generate
+    //       two callbacks.
+    [_completionLock lock];
+    
+    [ADURLProtocol unregisterProtocol];
+    
+    if ( _completionBlock )
+    {
+        void (^completionBlock)( ADAuthenticationError *, NSURL *) = _completionBlock;
+        _completionBlock = nil;
+        
+        dispatch_async( [ADAuthenticationSettings sharedInstance].dispatchQueue, ^{
+            completionBlock( error, url );
+        });
+    }
+    
+    [_completionLock unlock];
+}
+
+#pragma mark - ADAuthenticationDelegate
+
+- (BOOL)endWebAuthenticationWithError:(ADAuthenticationError*) error
+                                orURL:(NSURL*)endURL
+{
+    if ( nil != _authenticationViewController && nil != _parentController)
+    {
+        // Dismiss the authentication view and dispatch the completion block
+        [_parentController dismissViewControllerAnimated:YES completion:^{
+            [self dispatchCompletionBlock:error URL:endURL];
+        }];
+    }
+    else if (nil != _authenticationWebViewController)
+    {
+        [_authenticationWebViewController stop];
+        [self dispatchCompletionBlock:error URL:endURL];
+    }
+    else
+    {
+        return NO;
+    }
+    
+    _parentController = nil;
+    _authenticationViewController    = nil;
+    _authenticationWebViewController = nil;
+    
+    return YES;
+}
+
+
+// The user cancelled authentication
+- (void)webAuthenticationDidCancel
+{
+    DebugLog();
+    
+    // Dispatch the completion block
+
+    ADAuthenticationError* error = [ADAuthenticationError errorFromCancellation];
+    [self endWebAuthenticationWithError:error orURL:nil];
+}
+
+// Authentication completed at the end URL
+- (void)webAuthenticationDidCompleteWithURL:(NSURL *)endURL
+{
+    DebugLog();
+    [self endWebAuthenticationWithError:nil orURL:endURL];
+}
+
+// Authentication failed somewhere
+- (void)webAuthenticationDidFailWithError:(NSError *)error
+{
+    // Dispatch the completion block
+    ADAuthenticationError* adError = [ADAuthenticationError errorFromNSError:error errorDetails:error.localizedDescription];
+    
+    [self endWebAuthenticationWithError:adError orURL:nil];
+}
+
+@end
+
+@implementation ADWebAuthController (Internal)
```

Comment:
Beginning of the category implementation. Code has been moved around but not changed.